### PR TITLE
[Infra] Check rule activation for each SC rule

### PR DESCRIPTION
**Pull request by quantumagi**

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.SmartContracts/Rules/CanGetSenderRule.cs [hunk 1]
```diff
@@ -6,22 +6,28 @@
...
 
 namespace Stratis.Bitcoin.Features.SmartContracts.Rules
 {
     public class CanGetSenderRule : UtxoStoreConsensusRule
     {
         private readonly ISenderRetriever senderRetriever;
+        private readonly ISmartContractActivationProvider smartContractActivationProvider;
 
-        public CanGetSenderRule(ISenderRetriever senderRetriever)
+        public CanGetSenderRule(ISenderRetriever senderRetriever, ISmartContractActivationProvider smartContractActivationProvider = null)
```
Comment (rowandh) [inline src/Stratis.Bitcoin.Features.SmartContracts/Rules/CanGetSenderRule.cs:19]: Is there any reason we'd ever want this to default to `null`?

File: src/Stratis.Bitcoin.Features.SmartContracts/PoS/SmartContractActivationProvider.cs [hunk 2]
```diff
@@ -52,5 +60,17 @@ public SmartContractPosActivationProvider(Network network, NodeDeployments nodeD
                 }
             };
         }
+
+        /// <inheritdoc/>
+        public bool SkipRule(RuleContext context)
```
Comment (rowandh) [inline src/Stratis.Bitcoin.Features.SmartContracts/PoS/SmartContractActivationProvider.cs:65]: Would be nice to see a few unit tests for this logic.

File: src/Stratis.Bitcoin.Features.SmartContracts/Rules/CanGetSenderRule.cs (hunk 1 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.SmartContracts/Rules/CanGetSenderRule.cs:19]: The smart contract activation provider is only used when a network does not have smart contracts activated from the start.

Comment (rowandh) [reply src/Stratis.Bitcoin.Features.SmartContracts/Rules/CanGetSenderRule.cs:19]: Right, but that logic would still be in the activation provider. On a new network the provider implementation could just return true.

File: src/Stratis.Bitcoin.Features.SmartContracts/PoS/SmartContractActivationProvider.cs (hunk 2 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.SmartContracts/PoS/SmartContractActivationProvider.cs:65]: Added.

File: src/Stratis.Bitcoin.Features.SmartContracts/Rules/CanGetSenderRule.cs (hunk 1 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Features.SmartContracts/Rules/CanGetSenderRule.cs:19]: The idea is to maintain backwards compatibility with existing unrelated networks that do not have an activation provider.  Activation providers can be added for other networks as separate PRs.

Comment (rowandh) [reply src/Stratis.Bitcoin.Features.SmartContracts/Rules/CanGetSenderRule.cs:19]: Sounds reasonable

File: src/Stratis.Bitcoin.Features.SmartContracts/PoS/SmartContractActivationProvider.cs (hunk 2 above)
Comment (rowandh) [reply src/Stratis.Bitcoin.Features.SmartContracts/PoS/SmartContractActivationProvider.cs:65]: Thanks!

**Review by rowandh — APPROVED** [review]
LGTM

**Pull request merged**